### PR TITLE
[GEOS-11306] API Fix for GetFeature lazy JDBC count(*) in Java 21 and Java 17

### PR DESCRIPTION
**Pull request by jodygarnett**
[GEOS-11306](https://osgeo-org.atlassian.net/browse/GEOS-11306) [Powered by Pull Request Badge](https://pullrequestbadge.com/?utm_medium=github&utm_source=geoserver&utm_campaign=badge_info)

This required an update to apache-commons-lang3 covered by [GEOS-11307].

This issue is with respect to operating in Java 17 and Java 21 environments; relying on GitHub workflows to check regressions here (rather than new test cases).
  

# Checklist

- [x] I have read the [contribution guidelines](https://github.com/geoserver/geoserver/blob/main/CONTRIBUTING.md).
- [x] I have sent a [Contribution Licence Agreement](https://docs.geoserver.org/latest/en/developer/policies/committing.html) (not required for small changes, e.g., fixing typos in documentation).
- [x] First PR targets the `main` branch (backports managed later; ignore for branch specific issues).
- [ ] All the build checks are green ([see automated QA checks](https://docs.geoserver.org/latest/en/developer/qa-guide/index.html)).

For core and extension modules:

- [ ] New unit tests have been added covering the changes.
- [ ] [Documentation](https://github.com/geoserver/geoserver/tree/main/doc/en/user/source) has been updated (if change is visible to end users).
- [ ] The [REST API docs](https://github.com/geoserver/geoserver/tree/main/doc/en/api/1.0.0) have been updated (when changing configuration objects or the REST controllers).
- [x] There is an issue in the [GeoServer Jira](https://osgeo-org.atlassian.net/browse/GEOS/summary) (except for changes that do not affect administrators or end users in any way).
- [x] Commit message(s) must be in the form ``[GEOS-XYZWV] Title of the Jira ticket``.
- [x] Bug fixes and small new features are presented as a single commit.
- [x] Each commit has a single objective (if there are multiple commits, each has a separate JIRA ticket describing its goal).

### CODE REVIEW

File: src/wfs/src/main/java/org/geoserver/wfs/GetFeature.java [hunk 1]
```diff
@@ -703,7 +703,11 @@ private BigInteger updateTotalCount(
             long total = getTotalCount(totalCountExecutors);
             totalCount = BigInteger.valueOf(total);
         } else {
-            if (SystemUtils.IS_JAVA_11 || SystemUtils.IS_JAVA_1_8) {
+            if (SystemUtils.IS_JAVA_21
+                    || SystemUtils.IS_JAVA_17
+                    || SystemUtils.IS_JAVA_11
+                    || SystemUtils.IS_JAVA_1_8) {
```
Comment (mprins) [inline src/wfs/src/main/java/org/geoserver/wfs/GetFeature.java:709]: Java 8 is no longer a supported platform

Comment (aaime) [reply src/wfs/src/main/java/org/geoserver/wfs/GetFeature.java:709]: That's a good point indeed. Perhaps this is more compact:

```
     if (SystemUtils.isJavaVersionAtMost(SystemUtils.IS_JAVA_21)
```

**Comment by jodygarnett** [comment]
Folks I am suspicious that both JDK 17 and JDK 21 build are failing, consistently in `gs-app-schema-test`.

I am starting to suspect that just updating to compatible library is not sufficient.

**Comment by aaime** [comment]
@jodygarnett we recently merged a couple of test fixes from @groldan, exactly in app-schema... is this branch up to date? Maybe you just need to rebase.

**Comment by jodygarnett** [comment]
I will rebase, however since the failure is to do with count receiving 0 (rather than 2) I expect it has to do with this change directly.

File: src/wfs/src/main/java/org/geoserver/wfs/GetFeature.java [hunk 2]
```diff
@@ -703,7 +703,8 @@ private BigInteger updateTotalCount(
             long total = getTotalCount(totalCountExecutors);
             totalCount = BigInteger.valueOf(total);
         } else {
-            if (SystemUtils.IS_JAVA_11 || SystemUtils.IS_JAVA_1_8) {
+            if (SystemUtils.isJavaVersionAtMost(SystemUtils.IS_JAVA_21)) {
```
Comment (aaime) [inline src/wfs/src/main/java/org/geoserver/wfs/GetFeature.java:706]:
```suggestion
            if (SystemUtils.isJavaVersionAtMost(JavaVersion.JAVA_21)) {
```

**Comment by jodygarnett** [comment]
Testing with Java 17, and we have an answer:

```
org.springframework.cglib.core.CodeGenerationException: java.lang.IllegalAccessException-->module java.base does not open java.math to unnamed module @1068e947
```

While GeoServer does not log this exception it prevents this approach from working.

**Comment by jodygarnett** [comment]
Adding ``--add-opens=java.base/java.math=ALL-UNNAMED`` does not help:
```
java.lang.IllegalArgumentException: $java.math.BigInteger$$EnhancerByCGLIB$$e4f5dd18 not in same package as lookup class
```

**Comment by jodygarnett** [comment]
I am pushing up some defensive programming here, but the issue of a performance regression on Java 17 remains open.

I think we will need to make an API change here to use Number (rather than BigInteger).

**Comment by jodygarnett** [comment]
Yeah, and the time of the release cycle is good for an API change *why does everything happen right before release candidate....*

**Comment by jodygarnett** [comment]
@aaime I have some possible funding to address this; but it is unlikely to make the release candidate deadline:

Option: I could make the API change to Number now, leaving room for future development.

Option: I could wait and ask for permission to make an API and backport due to how impactful the regression is.

The availability of "run in Java 11" mitigation does take some pressure off.

**Comment by aaime** [comment]
Random ideas here... the reason to have the proxy is that the number is not used in some output formats, so it's computed lazily. If you go for an API change, rather than Number it could be a Provider<BigInteger>, makes things explicit.

That could be done at the level of the wrapper, FeatureCollectionResponse, where you set the provider, and force the computation only when the property is actually requested (few formats do) or when unadapt over WFS 2.0 requests is called, which is done for GML only, and update the WFS 2.0 adaptee at that point (for WFS 1.1 the total number is not require). 
That would make for a more isolated change I believe?

**Comment by jodygarnett** [comment]
I like it, that sounds cleaner then just changing to Number, and being forced to implement my own Number class.

Good thinking Andrea!

**Comment by jodygarnett** [comment]
I really like the Supplier<BigInteger> change proposed; I do not have customer approval this week so risk requesting a API change backport. 

If anyone else wishes to pick this up and make the API change I do have capacity to assist / review / test.

**Comment by jodygarnett** [comment]
Jotting down some notes on how to implement delegating getTotalNumberOfFeatures() by default.
```
public FeatureCollectionResult extends RequestObject {
    Supplier<BigInteger> lazyTotalNumberOfFeatures = () -> getTotalNumberOfFeatures();
    public Supplier<BigInteger> getLazyTotalNumberOfFeatures() { ... }
    public void setLazyTotalNumberOfFeatures(Supplier<BigInteger>) { ... }
    public abstract BigInteger getTotalNumberOfFeatures();
    public void setNumberOfFeatures(BigInteger n;)
}
```

**Comment by aaime** [comment]
Yes. Mind you need to also update "unadapt" so that it sets the property in the EMF model, for the GML encoders to work properly.

**Comment by jodygarnett** [comment]
Gak, I have recieved the go ahead to do this API change this morning. Going to update this PR and ask for review.

I am considering this a pretty large performance regression so I hope it can be treated as a bug with respect to back porting.

**Comment by jodygarnett** [comment]
> Yes. Mind you need to also update "unadapt" so that it sets the property in the EMF model, for the GML encoders to work properly.

@aaime I did not see the need to do this as I have implemented setTotalCount(BigInteger) and getTotalCount() to respect the EMF model. Perhaps I am not understanding EObject workflow in sufficient detail.

File: src/wfs/src/main/java/org/geoserver/wfs/request/FeatureCollectionResponse.java [hunk 3]
```diff
@@ -53,15 +62,84 @@ public void setTimeStamp(Calendar timeStamp) {
...
     public abstract BigInteger getNumberOfFeatures();
 
+    /**
+     * Number of features included in response.
+     *
+     * @param n Number of features included in response
+     */
     public abstract void setNumberOfFeatures(BigInteger n);
 
-    public abstract BigInteger getTotalNumberOfFeatures();
```
Comment (aaime) [inline src/wfs/src/main/java/org/geoserver/wfs/request/FeatureCollectionResponse.java:62]: I'd keep the existing getter and setter based on the lazy supplier

File: src/wfs/src/main/java/org/geoserver/wfs/request/FeatureCollectionResponse.java [hunk 4]
```diff
@@ -53,15 +62,84 @@ public void setTimeStamp(Calendar timeStamp) {
...
+     * Used to calculate total number of features on demand (only if needed). This allows formats
+     * that do not need the total to avoid calculating this expensive result, it also may be that
+     * some data stores can better estimate this total is obtained after traversing results.
+     *
+     * @param totalNumberOfFeatures Delayed calculation of total number of featuers.
+     */
+    public void setLazyTotalNumberOfFeatures(Supplier<BigInteger> totalNumberOfFeatures) {
+        if (lazyTotalNumberOfFeatures != null) {
+            throw new IllegalStateException("Total number of features has already been set");
+        }
```
Comment (aaime) [inline src/wfs/src/main/java/org/geoserver/wfs/request/FeatureCollectionResponse.java:98]: No need to check this, especially if you remove the getter for the supplier, which does not seem to be required anyways.

**Review by aaime — DISMISSED** [review]

**Comment by aaime** [comment]
The backport to `2.23.x` failed:
```
The process '/usr/bin/git' failed with exit code 1
```
<details>
<summary>stderr</summary>

```
error: could not apply dc9094a3e1... [GEOS-11306] Java 17 does not support GetFeature lazy JDBC count(*)
hint: After resolving the conflicts, mark them with
hint: "git add/rm <pathspec>", then run
hint: "git cherry-pick --continue".
hint: You can instead skip this commit with "git cherry-pick --skip".
hint: To abort and get back to the state before "git cherry-pick",
hint: run "git cherry-pick --abort".

```
</details>

<details>
<summary>stdout</summary>

```
Auto-merging src/pom.xml
[backport-7428-to-2.23.x 062021fa02] [GEOS-11307] Update apache-commons-lang3 to version 3.14.0 for Java 17 and Java 21 support
 Author: Jody Garnett <jody.garnett@gmail.com>
 Date: Fri Feb 16 14:07:06 2024 -0800
 1 file changed, 1 insertion(+), 1 deletion(-)
Auto-merging src/wfs/src/main/java/org/geoserver/wfs/GetFeature.java
CONFLICT (content): Merge conflict in src/wfs/src/main/java/org/geoserver/wfs/GetFeature.java

```
</details>

To backport manually, run these commands in your terminal:
```bash
# Fetch latest updates from GitHub
git fetch
# Create a new working tree
git worktree add .worktrees/backport-2.23.x 2.23.x
# Navigate to the new working tree
cd .worktrees/backport-2.23.x
# Create a new branch
git switch --create backport-7428-to-2.23.x
# Cherry-pick the merged commit of this pull request and resolve the conflicts
git cherry-pick 6d77ecb96199190671939508606f96b26eca1853,dc9094a3e1110a1f6601bbee227b77616f1cacc6
# Push it to GitHub
git push --set-upstream origin backport-7428-to-2.23.x
# Go back to the original working tree
cd ../..
# Delete the working tree
git worktree remove .worktrees/backport-2.23.x
```
Then, create a pull request where the `base` branch is `2.23.x` and the `compare`/`head` branch is `backport-7428-to-2.23.x`.

**Comment by aaime** [comment]
@jodygarnett found the new lazy counter is not quite working as expected, it's true that it's delaying the count, but it's not saving the result. The output machinery can ask for the value multiple times, and as a result, it gets computed multiple times. 
Long story short, it made things better for output formats that do not need the count, but worse for those that do...

**Comment by klbjerg** [comment]
Could be improved by wrapping in Guava Supplier memoize( ) - see https://guava.dev/releases/31.0-jre/api/docs/com/google/common/base/Suppliers.html#memoize(com.google.common.base.Supplier)

**Comment by aaime** [comment]
We already have guava in the classpath, just FYI.

**Comment by KorewaLidesu** [comment]
> This PR does not address the issue, but does provide a debug message to the logs now to indicate the problem.
> 
> As such it could be accepted "as is", leaving the issue open.
> 
> (Of course if anyone knows how to open `java.math` to reflection and constructor call from a different package we may be able to address the issue with a documentation fix)

You can check out this project
https://github.com/Rongmario/ImagineBreaker

**Pull request merged**